test/spec.ml

```diff
@@ -0,0 +1,224 @@
+    rather tries to make the correspondance with the specification as clear as
+    possible. For this reason, this code should not be used in production. *)
+
+let ( ++ ) = ( ^ )
+let list prehash = List.fold_left (fun acc x -> acc ++ prehash x) ""
+
+(** {2 Blake2b} *)
+
+type hash = string
+
+module Blake2b = Digestif.Make_BLAKE2B (struct
+  let digest_size = 32
+end)
+
+let blake2b s = Blake2b.digest_string s |> Blake2b.to_raw_string
+
+(** {2 Integers}
+
+    See section `Integers` of SPEC.md *)
+
+let fixed_int64 i =
+  let buf = Buffer.create 8 in
+  Buffer.add_int64_be buf i;
+  Buffer.contents buf
+
+let fixed_int i = Int64.of_int i |> fixed_int64
+
+let leb128_int i =
+  (* The final size of the result may be smaller (but not greater) than 8. *)
+  let buf = Buffer.create 8 in
+  let rec loop i =
+    let b = i land 127 in
+    let i = i lsr 7 in
+    (if i <> 0 then b lor 128 else b) |> Buffer.add_uint8 buf;
+    if i = 0 then Buffer.contents buf else loop i
+  in
+  loop i
+
+(** {2 Contents}
+
+    See section `Contents` of SPEC.md *)
+
+let content c = fixed_int (Bytes.length c) ++ Bytes.unsafe_to_string c
+
+(** {2 Commits}
+
+    See the section `Commits` of SPEC.md *)
+
+type commit_metadata = { date : int64; author : string; message : string }
+
+let commit_metadata { date; author; message } =
+  fixed_int64 date
+  ++ fixed_int (String.length author)
+  ++ author
+  ++ fixed_int (String.length message)
+  ++ message
+
+type commit = { tree : hash; parents : hash list; metadata : commit_metadata }
+
+let commit { tree; parents; metadata } =
+  fixed_int 32
+  ++ tree
+  ++ fixed_int (List.length parents)
+  ++ list (fun h -> fixed_int 32 ++ h) (List.sort String.compare parents)
+  ++ commit_metadata metadata
+
+(** {2 Trees}
+
+    See section `Trees` of SPEC.md *)
+
+type entry_kind = Content | Node
+type tree_entry = { name : string; kind : entry_kind; hash : hash }
+
+(** {3 Nodes}
+
+    When a tree contains at most 256 entries, it is encoded as a flat list of
+    entries
+
+    See section `Nodes` of SPEC.md *)
+
+type node_tree = tree_entry list
+
+let node_entry { name; kind; hash } =
+  (match kind with
+  | Content -> "\255\000\000\000\000\000\000\000"
+  | Node -> "\000\000\000\000\000\000\000\000")
+  ++ leb128_int (String.length name)
+  ++ name
+  ++ fixed_int 32
+  ++ hash
+
+let node t =
+  assert (List.length t <= 256);
+  fixed_int (List.length t)
+  ++ list node_entry (List.sort (fun e e' -> String.compare e.name e'.name) t)
+
+(** {3 Inodes}
+
+    Trees that contain more than 256 entries are first transforned into inodes,
+    before their final encoding.
+
+    See section `Inodes values` of SPEC.md *)
+
+let inode_entry { name; kind; hash } =
+  leb128_int (String.length name)
+  ++ name
+  ++ (match kind with Content -> "\001" | Node -> "\000")
+  ++ hash
+
+type inode_value = tree_entry list
+
+let inode_value v =
+  assert (List.length v <= 32);
+  "\000"
+  ++ String.make 1 (Char.chr (List.length v))
```

Comment:
why not `leb128_int  (List.length v)` ?

Reply:
That's how it's written in the spec :)
I agree it's odd and I'm very keen to change both to use LEB.
(same for other occurrences)

Reply:
I think it's written like this in the spec because `leb128 n` is the same as `\n` is `n <= 128` if I recall correctly so when `leb128 n` is 1 char size the leb128 is hidden.

Reply:
if it's clearer, maybe add a `val short_int: int -> string` and use this for short ints.

Reply:
(with an `assert (n < 128)` in there to make sure it's used properly.

Reply:
I think using leb128 everywhere in the inode implementation is actually a good thing to show that you only need leb128 to encode integers for inodes

Reply:
Agreed, I fixed that in c34c40b96711d13aba174d323d7c68856644a2c3.